From 450880b14af19d78528fd0308bd2445624f0ee32 Mon Sep 17 00:00:00 2001
From: Tony Wickham <twickham@google.com>
Date: Fri, 22 Aug 2025 14:49:52 -0700
Subject: [PATCH] Add onKeyEvent, to support KEYCODE_HOME

- Also added displayId to onBackEvent

Bug: 381339822
Flag: com.android.launcher3.home_button_uses_keycode_home
Test: Manual, OverviewCommandHelperTest
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:622a882e1f17d108a336ecfd3e4eb7aa9707f1e6)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:7ff1e38f7d707d860d9bc7557056050c4aaf4025)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:ec15cee125b81d66eee0a24682b223c1f8713833)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:195f8b0e4229ceb2b56ca66b2507f80f25ef1062)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:920c6db2d3249ec03c80ce1ba0fbf2553ad6aa62)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:4bb602c9721381fb1495f74fbe70c4c763e26cc3)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:e52f0e2d5ca8c7daff42ae49919386c7fc51b725)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:1679e88ef0a5f26769a6927725f14b185e6d91dc)
Merged-In: I55e2ee7adfa40c9d8f289a2b4c76aa12f80e78c3
Change-Id: I55e2ee7adfa40c9d8f289a2b4c76aa12f80e78c3
---
 .../systemui/shared/recents/ISystemUiProxy.aidl       |  9 +++++++--
 .../systemui/recents/OverviewProxyService.java        | 11 +++++++++++
 2 files changed, 18 insertions(+), 2 deletions(-)

diff --git a/packages/SystemUI/shared/src/com/android/systemui/shared/recents/ISystemUiProxy.aidl b/packages/SystemUI/shared/src/com/android/systemui/shared/recents/ISystemUiProxy.aidl
index 1c532fe7a529..ecb47d03c004 100644
--- a/packages/SystemUI/shared/src/com/android/systemui/shared/recents/ISystemUiProxy.aidl
+++ b/packages/SystemUI/shared/src/com/android/systemui/shared/recents/ISystemUiProxy.aidl
@@ -98,7 +98,7 @@ interface ISystemUiProxy {
     void expandNotificationPanel() = 29;
 
     /**
-     * Notifies SystemUI to invoke Back.
+     * Notifies SystemUI of KEYCODE_BACK, sending a ACTION_DOWN followed by ACTION_UP.
      */
     void onBackPressed() = 44;
 
@@ -125,5 +125,10 @@ interface ISystemUiProxy {
      */
     void toggleNotificationPanel() = 50;
 
-    // Next id = 51
+    /**
+     * Notifies SystemUI of a KeyEvent of the specified type (e.g. KEYCODE_BACK, KEYCODE_HOME).
+     */
+    oneway void onKeyEvent(int keycode) = 60;
+
+    // Next id = 61
 }
diff --git a/packages/SystemUI/src/com/android/systemui/recents/OverviewProxyService.java b/packages/SystemUI/src/com/android/systemui/recents/OverviewProxyService.java
index 4d005bebd99e..ec2b8bb21243 100644
--- a/packages/SystemUI/src/com/android/systemui/recents/OverviewProxyService.java
+++ b/packages/SystemUI/src/com/android/systemui/recents/OverviewProxyService.java
@@ -18,6 +18,7 @@ package com.android.systemui.recents;
 
 import static android.content.pm.PackageManager.MATCH_SYSTEM_ONLY;
 import static android.view.Display.DEFAULT_DISPLAY;
+import static android.view.KeyEvent.KEYCODE_BACK;
 import static android.view.MotionEvent.ACTION_CANCEL;
 import static android.view.MotionEvent.ACTION_DOWN;
 import static android.view.MotionEvent.ACTION_UP;
@@ -248,6 +249,16 @@ public class OverviewProxyService implements CallbackController<OverviewProxyLis
                     onTaskbarStatusUpdated(visible, stashed));
         }
 
+        @Override
+        public void onKeyEvent(int keycode) {
+            verifyCallerAndClearCallingIdentityPostMain(
+                    "onKeyEvent " + KeyEvent.keyCodeToString(keycode),
+                    () -> {
+                        sendEvent(KeyEvent.ACTION_DOWN, keycode);
+                        sendEvent(KeyEvent.ACTION_UP, keycode);
+                    });
+        }
+
         @Override
         public void notifyTaskbarAutohideSuspend(boolean suspend) {
             verifyCallerAndClearCallingIdentityPostMain("notifyTaskbarAutohideSuspend", () ->
-- 
2.34.1

